File: frontend/admin/src/api/system/base_area.ts
import service from "@/utils/request";
import {
  type CreateBaseAreaRequest,
  type DeleteBaseAreaRequest,
  type GetBaseAreaRequest,
  type TreeBaseAreaRequest,
  type TreeBaseAreaResponse,
  type OptionBaseAreaRequest,
  type SetBaseAreaStatusRequest,
  type BaseAreaForm,
  type BaseAreaService,
  type UpdateBaseAreaRequest
} from "@/rpc/system/admin/v1/base_area";
import type { Empty } from "@/rpc/google/protobuf/empty";
import type { TreeOptionResponse } from "@/rpc/common/v1/common";

const BASE_AREA_URL = "/v1/admin/base/area";

/** 行政区域服务。 */
export class BaseAreaServiceImpl implements BaseAreaService {

  /** 查询树形选择 */
  OptionBaseArea(request?: OptionBaseAreaRequest): Promise<TreeOptionResponse> {
    return service<OptionBaseAreaRequest, TreeOptionResponse>({
      url: BASE_AREA_URL + "/option",
      method: "get",
      params: request
    });
  }

  /** 查询树形列表 */
  TreeBaseArea(request: TreeBaseAreaRequest): Promise<TreeBaseAreaResponse> {
    return service<TreeBaseAreaRequest, TreeBaseAreaResponse>({
      url: BASE_AREA_URL + "/tree",
      method: "get",
      params: request
    });
  }

  /** 查询详情 */
  GetBaseArea(request: GetBaseAreaRequest): Promise<BaseAreaForm> {
    return service<GetBaseAreaRequest, BaseAreaForm>({
      url: BASE_AREA_URL + "/" + request.id,
      method: "get"
    });
  }

  /** 创建 */
  CreateBaseArea(request: CreateBaseAreaRequest): Promise<Empty> {
    return service<BaseAreaForm | undefined, Empty>({
      url: BASE_AREA_URL,
      method: "post",
      data: request.base_area
    });
  }

  /** 更新 */
  UpdateBaseArea(request: UpdateBaseAreaRequest): Promise<Empty> {
    return service<BaseAreaForm | undefined, Empty>({
      url: BASE_AREA_URL + "/" + request.id,
      method: "put",
      data: request.base_area
    });
  }

  /** 删除 */
  DeleteBaseArea(request: DeleteBaseAreaRequest): Promise<Empty> {
    return service<DeleteBaseAreaRequest, Empty>({
      url: BASE_AREA_URL + "/" + request.ids,
      method: "delete"
    });
  }

  /** 设置状态 */
  SetBaseAreaStatus(request: SetBaseAreaStatusRequest): Promise<Empty> {
    return service<SetBaseAreaStatusRequest, Empty>({
      url: BASE_AREA_URL + "/" + request.id + "/status",
      method: "put",
      data: request
    });
  }
}

export const defBaseAreaService = new BaseAreaServiceImpl();

File: backend/service/system/admin/biz/base_area.go
package biz

import (
	"context"

	commonv1 "shop/api/gen/go/common/v1"
	systemadminv1 "shop/api/gen/go/system/admin/v1"

	"shop/pkg/biz"
	"shop/pkg/errorsx"
	"shop/pkg/gen/data"
	"shop/pkg/gen/models"

	"github.com/liujitcn/go-utils/mapper"
	_string "github.com/liujitcn/go-utils/string"
	"github.com/liujitcn/gorm-kit/repository"
)

// BaseAreaCase 行政区域业务实例。
type BaseAreaCase struct {
	*biz.BaseCase
	*data.BaseAreaRepository
	formMapper *mapper.CopierMapper[systemadminv1.BaseAreaForm, models.BaseArea]
	mapper     *mapper.CopierMapper[systemadminv1.BaseArea, models.BaseArea]
}

// NewBaseAreaCase 创建行政区域业务实例。
func NewBaseAreaCase(baseCase *biz.BaseCase, baseAreaRepo *data.BaseAreaRepository) *BaseAreaCase {
	return &BaseAreaCase{
		BaseCase:           baseCase,
		BaseAreaRepository: baseAreaRepo,
		formMapper:         mapper.NewCopierMapper[systemadminv1.BaseAreaForm, models.BaseArea](),
		mapper:             mapper.NewCopierMapper[systemadminv1.BaseArea, models.BaseArea](),
	}
}

// OptionBaseArea 查询行政区域树形选择。
func (c *BaseAreaCase) OptionBaseArea(ctx context.Context, req *systemadminv1.OptionBaseAreaRequest) (*commonv1.TreeOptionResponse, error) {
	query := c.Query(ctx).BaseArea

	opts := make([]repository.QueryOption, 0, 2)
	opts = append(opts, repository.Order(query.ID.Asc()))
	if req.GetLazy() {
		opts = append(opts, repository.Where(query.ParentID.Eq(req.GetParentId())))
	}

	list, err := c.List(ctx, opts...)
	if err != nil {
		return nil, err
	}
	var hasChildren map[int64]struct{}
	hasChildren, err = c.listBaseAreaParentIDsWithChildren(ctx, list)
	if err != nil {
		return nil, err
	}
	return &commonv1.TreeOptionResponse{List: c.buildOptionBaseAreaOption(list, req.GetParentId(), req.GetLazy(), hasChildren)}, nil
}

// TreeBaseArea 查询行政区域树形列表。
func (c *BaseAreaCase) TreeBaseArea(ctx context.Context, req *systemadminv1.TreeBaseAreaRequest) (*systemadminv1.TreeBaseAreaResponse, error) {
	query := c.Query(ctx).BaseArea
	opts := make([]repository.QueryOption, 0, 3)
	opts = append(opts, repository.Order(query.ID.Asc()))
	// 搜索时跨层级匹配，避免懒加载树无法检索未展开节点。
	if req.GetName() != "" {
		opts = append(opts, repository.Where(query.Name.Like("%"+req.GetName()+"%")))
	} else {
		opts = append(opts, repository.Where(query.ParentID.Eq(req.GetParentId())))
	}

	list, err := c.List(ctx, opts...)
	if err != nil {
		return nil, err
	}
	var hasChildren map[int64]struct{}
	hasChildren, err = c.listBaseAreaParentIDsWithChildren(ctx, list)
	if err != nil {
		return nil, err
	}
	baseAreas := make([]*systemadminv1.BaseArea, 0, len(list))
	for _, item := range list {
		baseArea := c.mapper.ToDTO(item)
		_, baseArea.HasChildren = hasChildren[item.ID]
		baseAreas = append(baseAreas, baseArea)
	}
	return &systemadminv1.TreeBaseAreaResponse{BaseAreas: baseAreas}, nil
}

// GetBaseArea 查询行政区域详情。
func (c *BaseAreaCase) GetBaseArea(ctx context.Context, id int64) (*systemadminv1.BaseAreaForm, error) {
	baseArea, err := c.FindByID(ctx, id)
	if err != nil {
		return nil, err
	}
	return c.formMapper.ToDTO(baseArea), nil
}

// CreateBaseArea 创建行政区域。
func (c *BaseAreaCase) CreateBaseArea(ctx context.Context, req *systemadminv1.BaseAreaForm) error {
	baseArea := c.formMapper.ToEntity(req)
	return c.Create(ctx, baseArea)
}

// UpdateBaseArea 更新行政区域。
func (c *BaseAreaCase) UpdateBaseArea(ctx context.Context, id int64, req *systemadminv1.BaseAreaForm) error {
	baseArea := c.formMapper.ToEntity(req)
	baseArea.ID = id
	return c.UpdateByID(ctx, baseArea)
}

// DeleteBaseArea 删除行政区域。
func (c *BaseAreaCase) DeleteBaseArea(ctx context.Context, ids string) error {
	idList := _string.ConvertStringToInt64Array(ids)
	if len(idList) == 0 {
		return nil
	}
	query := c.Query(ctx).BaseArea
	for _, parentID := range idList {
		count, err := c.Count(ctx, repository.Where(query.ParentID.Eq(parentID)))
		if err != nil {
			return err
		}
		if count > 0 {
			return errorsx.HasChildrenConflict("删除行政区域失败，下面有行政区域", "base_area", "base_area")
		}
	}
	return c.DeleteByIDs(ctx, idList)
}

// buildOptionBaseAreaOption 构建行政区域树形选择。
func (c *BaseAreaCase) buildOptionBaseAreaOption(
	list []*models.BaseArea,
	parentID int64,
	lazy bool,
	hasChildren map[int64]struct{},
) []*commonv1.TreeOptionResponse_Option {
	res := make([]*commonv1.TreeOptionResponse_Option, 0)
	for _, item := range list {
		if int64(item.ParentID) != parentID {
			continue
		}
		option := &commonv1.TreeOptionResponse_Option{Label: item.Name, Value: int64(item.ID)}
		if lazy {
			_, option.HasChildren = hasChildren[item.ID]
		} else {
			option.Children = c.buildOptionBaseAreaOption(list, int64(item.ID), false, hasChildren)
		}
		res = append(res, option)
	}
	return res
}

// listBaseAreaParentIDsWithChildren 查询存在子节点的区域父级编号。
func (c *BaseAreaCase) listBaseAreaParentIDsWithChildren(ctx context.Context, list []*models.BaseArea) (map[int64]struct{}, error) {
	parentIDs := make([]int64, 0, len(list))
	for _, item := range list {
		parentIDs = append(parentIDs, item.ID)
	}
	hasChildren := make(map[int64]struct{}, len(parentIDs))
	if len(parentIDs) == 0 {
		return hasChildren, nil
	}

	query := c.Query(ctx).BaseArea
	childList, err := c.List(ctx, repository.Where(query.ParentID.In(parentIDs...)))
	if err != nil {
		return nil, err
	}
	for _, item := range childList {
		hasChildren[item.ParentID] = struct{}{}
	}
	return hasChildren, nil
}
